### ISSUE TITLE
Add an option to 'semi' to enforce semicolons except when in single line functions

### ISSUE THREAD

**Issue by shanewho** [eslint/eslint#4651]
With arrow functions, this won't be an issue but for us ES5 developers who like semicolons, it would be nice to not require them in single line, lambda-like function statements. Here is an example:

With Semicolons:
```javascript
[1,2,3,4,5].sort(function(a,b){ return a < b; });

[1,2,3,4,5]
    .sort(function(a,b){ return a < b; })
    .filter(function(i){ return i < 4; })
    .map(function(i){ return {value: i}; })
    .forEach(function(i){ console.log(i); });
```

Without Semicolons:
```javascript
[1,2,3,4,5].sort(function(a,b){ return a < b });

[1,2,3,4,5]
    .sort(function(a,b){ return a < b })
    .filter(function(i){ return i < 4 })
    .map(function(i){ return {value: i} })
    .forEach(function(i){ console.log(i) });
```

I wouldn't mind doing this work but would like to know if this would be wanted, what the rule option name/value should be, any potential issues/thoughts/etc.

**Comment by eslintbot** [comment]
@shanewho Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by nzakas** [comment]
Would this be for just one line functions? What about one line blocks (`if`, for example)?

**Comment by shanewho** [comment]
I guess you could generalize it to 'no semi required for single statement blocks' or something like that? 

Some possible examples:
```javascript
if(true) { console.log('TRUE') }
if(true) { console.log('TRUE'); }
if(true) console.log('TRUE');

while(i < 5) { i++ }
while(i < 5) { i++; }
while(i < 5) i++;

for(var i = 0; i < 5; i++) { console.log(i) }
for(var i = 0; i < 5; i++) { console.log(i); }
for(var i = 0; i < 5; i++) console.log(i);
```

Personally I write single line block statements without the blocks, so I don't really have an opinion on whether or not this **should** apply to if/while/for, but it definitely **could**. The difference between if/while/for and function is that function *requires* the braces and the others don't, but if you have an eslint rule that requires braces and don't want semi's in there, it could make sense.

**Comment by michaelficarra** [comment]
@shanewho Do you write single line try/catch without a semicolon?

**Comment by shanewho** [comment]
Like this?

```javascript
try { throw 'abc' } catch(e) { console.log(e) }

try { throw 'abc' } 
catch(e) { console.log(e) }
```

I don't, but I won't say it's a bad thing or that nobody does/shouldn't. I'm mainly interested in functions that work like lambdas (single statement with return passed as a parameter). 

That said, if a rule allowed no semi in lambda-like functions **and** it also allowed no semi in if/while/for/try/catch, I wouldn't mind a bit.

**Comment by nzakas** [comment]
The thing I'm trying to avoid is creating an option that is so specific that it can only ever be used by you. :) Adding an option that is specifically for functions on one line is really specific, whereas saying something like the last statement of a single-line block shouldn't have a semicolon. So the option would disallow semicolons in those spots always...not sure if that solves your case.

**Comment by ilyavolodin** [comment]
I'm not strongly opposed to creating an option like that, but at the same time, I feel that if you want to use semi-columns you should use them everywhere where they are needed. If you are relying on AST to clean up after you then you shouldn't use semicolumns at all. This in-between style can lead to newer people being confused and learning wrong things about the language when they read through your code. So I'm not sure I would want to promote such style through linter.

**Comment by shanewho** [comment]
@nzakas That would be just fine for me. Any of these seem ok to me:
- Required but allow no semi if it is the only statement in a block (my favorite...)
- Required but disallow for the last statement of a single line block  
(your suggestion, I think. Would work for me no problem, but seems a little weird to say it is required in case X but disallowed in case Y)
- Required but allow no semi if it is the only statement in a single line function (my case exactly, maybe too specific)
- Probably other options that would be fine by me as well


@ilyavolodin I see what you are saying but there are lots of times with filter/map/sort when you have a simple function that isn't a set of instructions so much as a simple return statement. The extra semi looks funny and is unnecessary (in my opinion).

Very common:
```
var userIds = users.map(function(user){return user.id;}); //Not great, in my opinion
var userIds = users.map(user=>user.id); //This would be ideal but requires ES6. Note there is no semi here.
var userIds = users.map(function(user){return user.id}); //The closest thing to lambda in es5
```

**Comment by mysticatea** [comment]
Is this the same as #4385?

**Comment by shanewho** [comment]
Yep, looks like it.

**Comment by nzakas** [comment]
Ooh good catch. Yeah, lets close in favor of #4385.
